README.md
# MongoDb-class

This [PHP class](mongodbclass.php) uses the [new MongoDb driver](https://www.php.net/manual/en/set.mongodb.php) to manage [MongoDb servers](https://www.mongodb.com/).
This repository is also a collection of resources to [learn and use MongoDb](documentation/learn.md).  
A [PHP example](example.php) file show how to use the class.  
This class only have the commonly used features with as few options as possible to keep things simple.  

# Quick information
* License: GNU [GPL-3.0-or-later](LICENSE.md)
* Releases: [CHANGELOG](CHANGELOG.md)
* Status: Working
* [Documentation](documentation/learn.md)

# License
This documentation is copyright (C) 2021 Peter Lembke.  
Permission is granted to copy, distribute and/or modify this document under the terms of the GNU Free Documentation License, Version 1.3 or any later version published by the Free Software Foundation; with no Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  
You should have received a copy of the GNU Free Documentation License along with this documentation. If not, see [https://www.gnu.org/licenses/](https://www.gnu.org/licenses/).  SPDX-License-Identifier: GFDL-1.3-or-later  


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# [1.0.0] - 2021-02-07

* [Github release notes](https://github.com/peterlembke/mongodb-class/releases/tag/v1.0.0)

Basic features in the mongodbclass.php. The working example.php. Documentation, PHPDOC. 


mongodb.php
<?php
/**
 * MongoDb class that use the new MongoDb driver
 *
 * @package mongo
 * @subpackage class
 */

/**
 * MongoDb class that use the new MongoDb driver
 *
 * @author      Peter Lembke <info@charzam.com>
 * @version     2021-02-06
 * @since       2021-02-06
 * @copyright   Copyright (c) 2021, Peter Lembke
 * @license     https://opensource.org/licenses/gpl-license.php GPL-3.0-or-later
 * @see         https://github.com/peterlembke/mongodb-class
 */
class mongodb
{
    /** @var string $host */
    protected $host = '';

    /** @var string $port */
    protected $port = 27017;

    /** @var string $userName */
    protected $userName = '';

    /** @var string $password */
    protected $password = '';

    /** @var string $databaseName */
    protected $databaseName = '';

    /** @var string $collectionName */
    protected $collectionName = '';

    /** @var MongoDB\Driver\Manager $manager */
    protected $manager = '';

    /** @var array $bulk */
    protected $bulk = [];

    /** @var array $bulkIndex */
    protected $bulkIndex = [];

    /**
     * Set the host.
     * Can be a domain name or an ip number
     * @param string $host
     */
    public function setHost(
        string $host = ''
    ): void {
        $this->host = $host;
    }

    /**
     * Set the host.
     * Can be a domain name or an ip number
     * @param int $port
     */
    public function setPort(
        int $port = 0
    ): void {
        $this->port = $port;
    }

    /**
     * Set the user name
     * @param string $userName
     */
    public function setUserName(
        string $userName = ''
    ): void {
        $this->userName = $userName;
    }

    /**
     * Set the password
     * @param string $password
     */
    public function setPassword(
        string $password = ''
    ): void {
        $this->password = $password;
    }

    /**
     * Set the database name
     * @param string $databaseName
     */
    public function setDatabaseName(
        string $databaseName = ''
    ): void {
        $this->databaseName = $databaseName;
    }

    /**
     * Set the collection name
     * @param string $collectionName
     */
    public function setCollectionName(
        string $collectionName = ''
    ): void {
        $this->collectionName = $collectionName;
    }

    /**
     * Connect to MongoDb using the credentials you have provided above
     * @return bool
     */
    public function connect(): bool {
        $user = [
            "username" => $this->userName,
            "password" => $this->password
        ];
        $connectionString = 'mongodb://' . $this->host . ':' . $this->port;

        try {
            $this->manager = new MongoDB\Driver\Manager($connectionString, $user);
        } catch (Exception $e) {
            return false;
        }
        return true;
    }

    /**
     * Get a string with databaseName.collectionName
     * @return string
     */
    protected function getDatabaseCollectionName(): string {
        return $this->databaseName . '.' . $this->collectionName;
    }

    /**
     * Execute a command and return the response array
     * @see https://www.php.net/manual/en/class.mongodb-driver-command.php
     * @see https://www.php.net/manual/en/mongodb-driver-manager.executecommand.php
     * @see https://www.php.net/manual/en/class.mongodb-driver-cursor.php
     * @param array $commandArray
     * @return array
     */
    protected function execute(array $commandArray = []): array {
        $command = new MongoDB\Driver\Command($commandArray);

        /** @var MongoDB\Driver\Cursor $cursor */
        $cursor = $this->manager->executeCommand($this->databaseName, $command);

        $responseArray = current($cursor->toArray())->values;

        return $responseArray;
    }

    /**
     * Execute a command and return the response array
     * @param array $filter
     * @param array $options
     * @return array
     */
    protected function query(array $filter = [], array $options = []): array {
        $query = new \MongoDB\Driver\Query($filter, $options);
        $rows = $this->manager->executeQuery($this->getDatabaseCollectionName(), $query);
        $resposeArray = $rows->toArray();

        return $resposeArray;
    }

    /**
     * Get an id object from an id string
     * @param string $id
     * @return \MongoDB\BSON\ObjectId
     */
    protected function getObjectId(string $id = ''): \MongoDB\BSON\ObjectId {
        return new \MongoDB\BSON\ObjectId($id);
    }

    /**
     * Clear the bulk
     */
    public function bulkClear(
    ): void {
        $this->bulk = [];
    }

    /**
     * Add an item to insert
     * @see https://www.php.net/manual/en/mongodb-driver-bulkwrite.insert.php
     * @param array $item
     */
    public function bulkInsert(
        array $item = []
    ): void {
        $this->bulk['insert'][] = $item;
    }

    /**
     * Add a command to delete. You must run bulkWrite
     * @see https://www.php.net/manual/en/mongodb-driver-bulkwrite.delete.php
     * @param array $filter
     * @param array $options
     */
    public function bulkDelete(
        array $filter = [],
        array $options = []
    ): void {
        $this->bulk['delete'][] = [$filter,$options];
    }

    /**
     * Add a command to update. You must run bulkWrite
     * @see https://www.php.net/manual/en/mongodb-driver-bulkwrite.update.php
     * @example $filter = ['name' => 'Peter'], $set = ['age' => 50, 'updated_at' => '2021-02-06']
     * @param array $filter
     * @param array $set
     */
    public function bulkUpdate(
        array $filter = [],
        array $set = []
    ): void {
        $options = ['$set' => $set]; // Set new values
        $this->bulk['update'][] = [$filter,$options];
    }

    /**
     * Run all your delete, insert, update
     * @return array
     */
    public function bulkWrite(): array {
        $bulk = new MongoDB\Driver\BulkWrite;
        foreach ($this->bulk as $commandTypeString => $commandCollectionArray) {

            switch ($commandTypeString) {
                case 'insert':
                    foreach ($commandCollectionArray as $item) {
                        $bulk->insert($item);
                    }
                    break;
                case 'update':
                    foreach ($commandCollectionArray as $item) {
                        $bulk->update($item[0], $item[1]);
                    }
                    break;
                case 'delete':
                    foreach ($commandCollectionArray as $item) {
                        $bulk->delete($item[0], $item[1]);
                    }
                    break;
            }
        }

        /** @var MongoDB\Driver\WriteResult $writeResult */
        $writeResult = $this->manager->executeBulkWrite($this->getDatabaseCollectionName(), $bulk);

        $resultArray = $writeResult->_toArray();

        return $resultArray;
    }

    /**
     * Load and return an item found by its id
     * @param string $itemId
     * @return array
     */
    public function loadById(string $itemId = ''): array {
        $objectId = $this->getObjectId($itemId);
        $filter = ['_id' => $objectId];
        $options = ['projection' => ['_id' => 0]];

        return $this->query($filter, $options);
    }

    /**
     * Load and return a collection array with items
     * @example $filter = ['foo' => 'yes'], $options = ['projection' => ['_id' => 0]]
     * @param array $filter
     * @param array $options
     * @return array
     */
    public function loadCollection(array $filter = [], array $options = []): array {
        return $this->query($filter, $options);
    }

    /**
     * Get statistics for the collection you have set
     * @return array
     */
    public function collectionStatistics(): array {
        $commandArray = ["collstats" => $this->collectionName];
        return $this->execute($commandArray);
    }

    /**
     * Add an index definition that you will create with bulkIndexCreate
     * @see https://stackoverflow.com/questions/46819481/php7-mongodb-create-index
     * @param string $indexName
     * @param array $keyArray
     * @param bool $unique
     * @param int $expireAfterSeconds
     */
    public function bulkIndexAdd(
        string $indexName = '',
        array $keyArray = [],
        bool $unique = false,
        int $expireAfterSeconds = 0
    ): void {

        $index = [
            "name" => $indexName,
            "key" => $keyArray,
            "ns" => $this->getDatabaseCollectionName(),
            'unique' => $unique
        ];

        if ($expireAfterSeconds > 0) {
            $index['expireAfterSeconds'] = $expireAfterSeconds;
        }

        $this->bulkIndex[] = $index;
    }

    /**
     * Create all indexes you have added with bulkIndexAdd
     * @see https://www.php.net/manual/en/class.mongodb-driver-cursor.php
     */
    public function bulkIndexCreate(): array {
        $commandArray = [
            "createIndexes" => $this->collectionName,
            "indexes" => $this->bulkIndex
        ];
        return $this->execute($commandArray);
    }

    /**
     * Delete an index
     * @param string $indexName
     * @return array
     */
    public function indexDelete(string $indexName = ''): array {
        $commandArray = [
            'dropIndexes' => $this->collectionName,
            'index' => $indexName
        ];
        return $this->execute($commandArray);
    }

    /**
     * Delete a database
     * @return array
     */
    public function databaseDelete(): array {
        $commandArray = ['dropDatabase' => 1];
        return $this->execute($commandArray);
    }

    /**
     * Delete a collection
     * @return array
     */
    public function collectionDelete(): array {
        $commandArray = ['drop' => $this->collectionName];
        return $this->execute($commandArray);
    }

    /**
     * Create a database
     */
    public function databaseCreate(): void {
        
    }

    /**
     * Get a list with all database names on the server
     * @param bool $nameOnly
     * @return array
     */
    public function databaseList(bool $nameOnly = true): array {
        $commandArray = ['listDatabases' => 1, 'nameOnly' => $nameOnly];
        return $this->execute($commandArray);
    }

    /**
     * Get a list with all collection names in the database
     * @param bool $nameOnly
     * @return array
     */
    public function collectionList(bool $nameOnly = true): array {
        $commandArray = ['listCollections' => 1, 'nameOnly' => $nameOnly];
        return $this->execute($commandArray);
    }

    /**
     * Create a collection
     * @see https://www.php.net/manual/en/class.mongodb-driver-command.php
     * @return array
     */
    public function collectionCreate(
    ): array {
        $commandArray = ['create' => $this->collectionName];
        return $this->execute($commandArray);
    }

    /**
     * Get information about the server
     * @see https://www.php.net/manual/en/mongodb-driver-command.construct.php
     * @return array
     */
    public function getBuildInformation(): array {
        $commandArray = ["buildinfo" => 1];
        return $this->execute($commandArray);
    }


}


mongodbclass.php
<?php
/**
 * MongoDb class that use the new MongoDb driver to perform common tasks with minimum of options.
 *
 * @author      Peter Lembke <info@charzam.com>
 * @version     2021-02-07
 * @since       2021-02-06
 * @copyright   Copyright (c) 2021, Peter Lembke
 * @license     https://opensource.org/licenses/gpl-license.php GPL-3.0-or-later
 * @see         https://github.com/peterlembke/mongodb-class
 */
class MongoDbClass
{
    /** @var string $host */
    protected $host = '';

    /** @var string $port */
    protected $port = 27017;

    /** @var string $userName */
    protected $userName = '';

    /** @var string $password */
    protected $password = '';

    /** @var string $databaseName */
    protected $databaseName = '';

    /** @var string $collectionName */
    protected $collectionName = '';

    /** @var MongoDB\Driver\Manager $manager */
    protected $manager = '';

    /** @var array $bulk */
    protected $bulk = [];

    /** @var array $bulkIndex */
    protected $bulkIndex = [];

    /**
     * Set the host.
     * Can be a domain name or an ip number
     * @param string $host
     */
    public function setHost(
        string $host = ''
    ): void {
        $this->host = $host;
    }

    /**
     * Set the host.
     * Can be a domain name or an ip number
     * @param int $port
     */
    public function setPort(
        int $port = 0
    ): void {
        $this->port = $port;
    }

    /**
     * Set the user name
     * @param string $userName
     */
    public function setUserName(
        string $userName = ''
    ): void {
        $this->userName = $userName;
    }

    /**
     * Set the password
     * @param string $password
     */
    public function setPassword(
        string $password = ''
    ): void {
        $this->password = $password;
    }

    /**
     * Set the database name
     * @param string $databaseName
     */
    public function setDatabaseName(
        string $databaseName = ''
    ): void {
        $this->databaseName = $databaseName;
    }

    /**
     * Set the collection name
     * @param string $collectionName
     */
    public function setCollectionName(
        string $collectionName = ''
    ): void {
        $this->collectionName = $collectionName;
    }

    /**
     * Connect to MongoDb using the credentials you have provided above
     * @return bool
     */
    public function connect(): bool {
        $user = [
            "username" => $this->userName,
            "password" => $this->password
        ];
        $connectionString = 'mongodb://' . $this->host . ':' . $this->port;

        try {
            $this->manager = new MongoDB\Driver\Manager($connectionString, $user);
        } catch (Exception $e) {
            return false;
        }
        return true;
    }

    /**
     * Get a string with databaseName.collectionName
     * @return string
     */
    public function getDatabaseCollectionName(): string {
        return $this->databaseName . '.' . $this->collectionName;
    }

    /**
     * Execute a command and return the response array
     * @see https://www.php.net/manual/en/class.mongodb-driver-command.php
     * @see https://www.php.net/manual/en/mongodb-driver-manager.executecommand.php
     * @see https://www.php.net/manual/en/class.mongodb-driver-cursor.php
     * @param array $commandArray
     * @param array $options
     * @return array
     */
    protected function execute(array $commandArray = []): array {
        $command = new MongoDB\Driver\Command($commandArray);

        $responseArray = [
            'answer' => false,
            'message' => 'Got nothing back from the command',
            'code_name' => ''
        ];

        try {
            /** @var MongoDB\Driver\Cursor $cursor */
            $cursor = $this->manager->executeCommand($this->databaseName, $command);
            $cursorArray = $cursor->toArray();
            if (count($cursorArray) > 0) {
                $responseArray = json_decode(json_encode($cursorArray), true);
            }
        } catch (MongoDb\Driver\Exception\CommandException $e) {
            $responseArray = [
                'answer' => false,
                'message' => $e->getMessage(),
                'code_name' => $e->getResultDocument()->codeName
            ];
        }

        return $responseArray;
    }

    /**
     * Execute a command and return the response array
     * @param array $filter
     * @param array $options
     * @return array
     */
    protected function query(array $filter = [], array $options = []): array {
        $query = new \MongoDB\Driver\Query($filter, $options);
        $rows = $this->manager->executeQuery($this->getDatabaseCollectionName(), $query);
        $resposeArray = $rows->toArray();

        return $resposeArray;
    }

    /**
     * Get an id object from an id string
     * @param string $id
     * @return \MongoDB\BSON\ObjectId
     */
    protected function getObjectId(string $id = ''): \MongoDB\BSON\ObjectId {
        return new \MongoDB\BSON\ObjectId($id);
    }

    /**
     * Clear the bulk
     */
    public function bulkClear(
    ): void {
        $this->bulk = [];
    }

    /**
     * Add an item to insert
     * @see https://www.php.net/manual/en/mongodb-driver-bulkwrite.insert.php
     * @param array $item
     */
    public function bulkInsert(
        array $item = []
    ): void {
        $this->bulk['insert'][] = $item;
    }

    /**
     * Add a command to delete. You must run bulkWrite
     * @see https://www.php.net/manual/en/mongodb-driver-bulkwrite.delete.php
     * @param array $filter
     * @param array $options
     */
    public function bulkDelete(
        array $filter = [],
        array $options = []
    ): void {
        $this->bulk['delete'][] = [$filter,$options];
    }

    /**
     * Add a command to update. You must run bulkWrite
     * @see https://www.php.net/manual/en/mongodb-driver-bulkwrite.update.php
     * @example $filter = ['name' => 'Peter'], $set = ['age' => 50, 'updated_at' => '2021-02-06']
     * @param array $filter
     * @param array $set
     */
    public function bulkUpdate(
        array $filter = [],
        array $set = []
    ): void {
        $options = ['$set' => $set]; // Set new values
        $this->bulk['update'][] = [$filter,$options];
    }

    /**
     * Run all your delete, insert, update
     * @return array
     */
    public function bulkWrite(): array {
        $bulk = new MongoDB\Driver\BulkWrite;
        foreach ($this->bulk as $commandTypeString => $commandCollectionArray) {

            switch ($commandTypeString) {
                case 'insert':
                    foreach ($commandCollectionArray as $item) {
                        $bulk->insert($item);
                    }
                    break;
                case 'update':
                    foreach ($commandCollectionArray as $item) {
                        $bulk->update($item[0], $item[1]);
                    }
                    break;
                case 'delete':
                    foreach ($commandCollectionArray as $item) {
                        $bulk->delete($item[0], $item[1]);
                    }
                    break;
            }
        }

        /** @var MongoDB\Driver\WriteResult $writeResult */
        $writeResult = $this->manager->executeBulkWrite($this->getDatabaseCollectionName(), $bulk);

        $resultArray = [
            'inserted_count' => $writeResult->getInsertedCount(),
            'updated_count' => $writeResult->getModifiedCount(),
            'deleted_count' => $writeResult->getDeletedCount()
        ];

        return $resultArray;
    }

    /**
     * Load and return an item found by its id
     * @param string $itemId
     * @return array
     */
    public function loadById(string $itemId = ''): array {
        $objectId = $this->getObjectId($itemId);
        $filter = ['_id' => $objectId];
        $options = []; // ['projection' => ['_id' => 0]];

        return $this->query($filter, $options);
    }

    /**
     * Load and return a collection array with items
     * @example $filter = ['foo' => 'yes'], $options = ['projection' => ['_id' => 0]]
     * @param array $filter
     * @param array $options
     * @return array
     */
    public function loadCollection(array $filter = [], array $options = []): array {
        return $this->query($filter, $options);
    }

    /**
     * Get statistics for the collection you have set
     * @param bool $compact
     * @return array
     */
    public function collectionStatistics(bool $compact = true): array {
        $commandArray = ["collstats" => $this->collectionName];
        $response = $this->execute($commandArray);
        if ($compact === false) {
            return $response;
        }
        $result = [];
        foreach ($response as $item) {
            unset($item['wiredTiger']);
            unset($item['indexDetails']);
            $name = $item['ns'];
            $result[$name] = $item;
        }
        return $result;
    }

    /**
     * Add an index definition that you will create with bulkIndexCreate
     * @see https://stackoverflow.com/questions/46819481/php7-mongodb-create-index
     * @param string $indexName
     * @param array $keyArray
     * @param bool $unique
     * @param int $expireAfterSeconds
     */
    public function bulkIndexAdd(
        string $indexName = '',
        array $keyArray = [],
        bool $unique = false,
        int $expireAfterSeconds = 0
    ): void {

        $index = [
            "name" => $indexName,
            "key" => $keyArray,
            "ns" => $this->getDatabaseCollectionName(),
            'unique' => $unique
        ];

        if ($expireAfterSeconds > 0) {
            $index['expireAfterSeconds'] = $expireAfterSeconds;
        }

        $this->bulkIndex[] = $index;
    }

    /**
     * Create all indexes you have added with bulkIndexAdd
     * @see https://www.php.net/manual/en/class.mongodb-driver-cursor.php
     */
    public function bulkIndexCreate(): array {
        $commandArray = [
            "createIndexes" => $this->collectionName,
            "indexes" => $this->bulkIndex
        ];
        return $this->execute($commandArray);
    }

    /**
     * Delete an index
     * @param string $indexName
     * @return array
     */
    public function indexDelete(string $indexName = ''): array {
        $commandArray = [
            'dropIndexes' => $this->collectionName,
            'index' => $indexName
        ];
        return $this->execute($commandArray);
    }

    /**
     * Delete an index
     * @param bool $nameOnly
     * @return array
     */
    public function indexList(bool $nameOnly = true): array {
        $commandArray = ['listIndexes' => $this->collectionName];
        $response = $this->execute($commandArray);
        if ($nameOnly === false) {
            return $response;
        }
        $result = [];
        foreach ($response as $item) {
            $result[] = $item['name'];
        }
        return $result;
    }

    /**
     * Get a list with all database names on the server
     * @param bool $nameOnly
     * @return array
     */
    public function databaseList(bool $nameOnly = true): array {
        $currentDatabase = $this->databaseName;
        $this->databaseName = 'admin';
        $commandArray = ['listDatabases' => 1, 'nameOnly' => $nameOnly];
        $response = $this->execute($commandArray);
        $this->databaseName = $currentDatabase;
        if ($nameOnly === false) {
            return $response;
        }
        $result = [];
        foreach ($response[0]['databases'] as $item) {
            $result[] = $item['name'];
        }
        return $result;
    }

    /**
     * Get a list with all collection names in the database
     * @param bool $nameOnly
     * @return array
     */
    public function collectionList(bool $nameOnly = true): array {
        $commandArray = ['listCollections' => 1, 'nameOnly' => $nameOnly];
        $response = $this->execute($commandArray);
        if ($nameOnly === false) {
            return $response;
        }
        $result = [];
        foreach ($response as $item) {
            $result[] = $item['name'];
        }
        return $result;
    }

    /**
     * Delete a database
     * @return array
     */
    public function databaseDelete(): array {
        $commandArray = ['dropDatabase' => 1];
        return $this->execute($commandArray);
    }

    /**
     * Delete a collection
     * @return array
     */
    public function collectionDelete(): array {
        $commandArray = ['drop' => $this->collectionName];
        return $this->execute($commandArray);
    }

    /**
     * Create a database
     * It seems to be enough to create a collection and then the database are also created
     */
    public function databaseCreate(): void {

    }

    /**
     * Create a collection
     * @see https://www.php.net/manual/en/class.mongodb-driver-command.php
     * @return array
     */
    public function collectionCreate(
    ): array {
        $commandArray = ['create' => $this->collectionName];
        return $this->execute($commandArray);
    }

    /**
     * Get information about the MongoDb software
     * @see https://www.php.net/manual/en/mongodb-driver-command.construct.php
     * @param bool $compact
     * @return array
     */
    public function getBuildInformation(bool $compact = true): array {
        $commandArray = ["buildinfo" => 1];
        $response = $this->execute($commandArray);
        return $response[0];
    }

    /**
     * You can send in a pipeline with stages and MQL commands
     * to get the data you want from the collection
     * @see https://www.php.net/manual/en/class.mongodb-driver-command.php#123163
     * @see https://docs.mongodb.com/manual/reference/operator/aggregation-pipeline/
     * @see https://docs.mongodb.com/manual/reference/command/aggregate/
     * @param array $pipeline
     * @return array
     */
    public function collectionAggregate(array $pipeline = []): array {
        $commandArray = [
            'aggregate' => $this->collectionName,
            'pipeline' => $pipeline,
            // 'explain' => true, // If aggregation do not work then uncomment to get more information
            'cursor' => new stdClass()
        ];
        return $this->execute($commandArray);
    }
}


example.php
<?php
/**
 * Example file that use the MongoDb class
 *
 * @author      Peter Lembke <info@charzam.com>
 * @version     2021-02-07
 * @since       2021-02-06
 * @copyright   Copyright (c) 2021, Peter Lembke
 * @license     https://opensource.org/licenses/gpl-license.php GPL-3.0-or-later
 * @see         https://github.com/peterlembke/mongodb-class
 */

include_once "mongodbclass.php";

class example
{
    /** @var MongoDbClass */
    protected $mongodb;

    public function __construct(
        MongoDbClass $mongodb
    ) {
        $this->mongodb = $mongodb;
    }

    public function run(): void {
        $this->connect();
        // $this->create();
        // $this->info();
        $this->query();
    }

    protected function connect(): void {
        $this->mongodb->setUserName('root');
        $this->mongodb->setPassword('infohub');
        $this->mongodb->setHost('172.19.0.4');
        $this->mongodb->setPort(27017);
        $this->mongodb->setDatabaseName('local');
        $this->mongodb->setCollectionName('example');
        $result = $this->mongodb->connect();

        if ($result === false) {
            exit('Can not connect to database');
        }
    }

    protected function create(): void {
        $this->mongodb->collectionCreate();
        $item = ['first_name' => 'Adam', 'last_name' => 'Adamson', 'city' => 'Stockholm', 'born' => 1922, 'telephone' => '555-1234', 'club' => 'aik'];
        $this->mongodb->bulkInsert($item);
        $item = ['first_name' => 'Bertil', 'last_name' => 'Bertilson', 'city' => 'Uppsala', 'born' => 1945, 'telephone' => '555-2345', 'club' => 'aik'];
        $this->mongodb->bulkInsert($item);
        $item = ['first_name' => 'Cesar', 'last_name' => 'Cesarson', 'city' => 'Uppsala', 'born' => 1951, 'telephone' => '555-3456', 'club' => 'aik'];
        $this->mongodb->bulkInsert($item);
        $item = ['first_name' => 'David', 'last_name' => 'Davidson', 'city' => 'Uppsala', 'born' => 1965, 'telephone' => '555-4567', 'club' => 'hif'];
        $this->mongodb->bulkInsert($item);
        $item = ['first_name' => 'Erik', 'last_name' => 'Erikson', 'city' => 'Stockholm', 'born' => 1971, 'telephone' => '555-5678', 'club' => 'hif'];
        $this->mongodb->bulkInsert($item);
        $this->mongodb->bulkWrite();

        $this->mongodb->bulkIndexAdd('city', ['city' => 1]);
        $this->mongodb->bulkIndexAdd('born', ['born' => 1]);
        $this->mongodb->bulkIndexAdd('club', ['club' => 1]);
        $this->mongodb->bulkIndexCreate();
    }

    protected function info(): void {
        $databaseCollectionName = $this->mongodb->getDatabaseCollectionName();
        $this->titlePrint('Current database and collection name');
        $this->prettyPrint([$databaseCollectionName]);

        $list = $this->mongodb->databaseList(true);
        $this->titlePrint('database list');
        $this->prettyPrint($list);

        $list = $this->mongodb->collectionList(true);
        $this->titlePrint('collection list');
        $this->prettyPrint($list);

        $list = $this->mongodb->indexList(true);
        $this->titlePrint('index list');
        $this->prettyPrint($list);

        $statistics = $this->mongodb->collectionStatistics();
        $this->titlePrint('collection statistics');
        $this->prettyPrint($statistics);

        $statistics = $this->mongodb->getBuildInformation();
        $this->titlePrint('build information');
        $this->prettyPrint($statistics);
    }

    protected function query(): void {
        // $this->loadById();
        // $this->loadBorn1922();
        // $this->loadGreaterThanAndIn();
        $this->aggregate();
    }

    /**
     * Each document has a unique id
     */
    protected function loadById(): void {
        $queryResult = $this->mongodb->loadById('601ecc8db556b81ba2793944');
        $this->titlePrint('load by id');
        $this->prettyPrint($queryResult);
    }

    /**
     * Example of a match of value. You can have more matching fields in the array
     */
    protected function loadBorn1922(): void {
        $filter = ['born' => 1922];
        $options = []; // ['projection' => ['_id' => 0]];
        $queryResult = $this->mongodb->loadCollection($filter, $options);
        $this->titlePrint('loadCollection born = 1922');
        $this->prettyPrint($queryResult);
    }

    /**
     * Example of $gt - greater than, and $in - in array
     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/
     * @example MQL {born: {$gt: 1950}, club: {$in: ['aik','hif']}}
     */
    protected function loadGreaterThanAndIn(): void {
        $filter = ['born' => ['$gt' => 1950], 'club' => ['$in' => ['aik', 'dif']]];
        $options = []; // ['projection' => ['_id' => 0]];
        $queryResult = $this->mongodb->loadCollection($filter, $options);
        $this->titlePrint('loadCollection born > 1950 & club = aik or dif');
        $this->prettyPrint($queryResult);
    }

    /**
     * Aggregation in a pipeline with stages is what MongoDb is about
     * This example show two stages
     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/group/
     * @see https://docs.mongodb.com/manual/reference/operator/aggregation/match/
     */
    protected function aggregate(): void {
        $pipeline = [
            [
                '$match' => [
                    'born' => ['$gt' => 1950],
                    'club' => ['$in' => ['hif', 'aik']]
                ]
            ],
            [
                '$group' => [
                    '_id' => '$club',
                    'club_count' => ['$sum' => 1]
                ]
            ]
        ];
        $queryResult = $this->mongodb->collectionAggregate($pipeline);
        $this->titlePrint('Aggregate a collection');
        $this->prettyPrint($queryResult);
    }

    protected function titlePrint(string $title = ''): void {
        echo '<h2>' . $title . '</h2>';
    }

    protected function prettyPrint(array $data = []): void {
        $jsonData = json_encode($data, JSON_PRETTY_PRINT);
        echo '<pre>' . $jsonData . '</pre>';
    }

}

$myMongoDbClass = new MongoDbClass();
$myExample = new example($myMongoDbClass);
$myExample->run();